File: tippmix.py
"""
Tippmix.hu odds fetcher.
Calls the public tippmix.hu JSON API directly with `requests` — no browser needed.
The list endpoints (best-games / last-minute) are POST; the per-event market
endpoint is GET. Maps Hungarian team names to English for the Poisson model.
"""
import requests
from concurrent.futures import ThreadPoolExecutor

API_BASE = "https://api.tippmix.hu"

# tippmix's API rejects requests without these (the list endpoints 404 otherwise).
_HEADERS = {
    "Accept": "application/json, text/plain, */*",
    "Content-Type": "application/json;charset=UTF-8",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko)",
    "Referer": "https://www.tippmix.hu/",
    "Accept-Language": "hu-HU",
}

# Hungarian name → English name mapping for WC 2026 teams
HU_TO_EN: dict[str, str] = {
    "Mexikó": "Mexico",
    "Dél-Afrika": "South Africa",
    "Marokkó": "Morocco",
    "Portugália": "Portugal",
    "Argentína": "Argentina",
    "Szaúd-Arábia": "Saudi Arabia",
    "Lengyelország": "Poland",
    "Németország": "Germany",
    "Japán": "Japan",
    "Spanyolország": "Spain",
    "Costa Rica": "Costa Rica",
    "Belgium": "Belgium",
    "Kanada": "Canada",
    "Horvátország": "Croatia",
    "Szerbia": "Serbia",
    "Brazília": "Brazil",
    "Svájc": "Switzerland",
    "Kamerun": "Cameroon",
    "Dél-Korea": "Korea Republic",
    "Ghána": "Ghana",
    "Uruguay": "Uruguay",
    "Franciaország": "France",
    "Ausztrália": "Australia",
    "Dánia": "Denmark",
    "Tunézia": "Tunisia",
    "Szenegál": "Senegal",
    "Hollandia": "Netherlands",
    "Ecuador": "Ecuador",
    "Anglia": "England",
    "Irán": "Iran",
    "Wales": "Wales",
    "Katar": "Qatar",
    "Egyesült Államok": "United States",
    "USA": "United States",
    "Panamá": "Panama",
    "Jamaika": "Jamaica",
    "Bolívia": "Bolivia",
    "Venezuela": "Venezuela",
    "Kolumbia": "Colombia",
    "Chile": "Chile",
    "Peru": "Peru",
    "Paraguay": "Paraguay",
    "Honduras": "Honduras",
    "Nigéria": "Nigeria",
    "Elefántcsontpart": "Ivory Coast",
    "Mali": "Mali",
    "Kenya": "Kenya",
    "Egyiptom": "Egypt",
    "DR Kongó": "DR Congo",
    "Algéria": "Algeria",
    "Szlovénia": "Slovenia",
    "Ausztria": "Austria",
    "Törökország": "Turkey",
    "Csehország": "Czech Republic",
    "Új-Zéland": "New Zealand",
    "Norvégia": "Norway",
    "Svédország": "Sweden",
    "Skócia": "Scotland",
    "Görögország": "Greece",
    "Románia": "Romania",
    "Ukrajna": "Ukraine",
    "Magyarország": "Hungary",
    "Olaszország": "Italy",
    "Bosznia-Hercegovina": "Bosnia & Herzegovina",
}


def _en(name: str) -> str:
    return HU_TO_EN.get(name, name)


def _parse_events(data: dict) -> list[dict]:
    """Extract 1X2 football matches from a tippmix API response."""
    matches = []
    for sport_block in data.get("data", []):
        if sport_block.get("sportId") != 1:
            continue
        for ev in sport_block.get("events", []):
            if ev.get("isLive") or ev.get("bettingStatus") != 0:
                continue
            participants = ev.get("eventParticipants", [])
            if len(participants) < 2:
                continue

            home_hu = participants[0]["participantName"]
            away_hu = participants[1]["participantName"]
            home_en = _en(home_hu)
            away_en = _en(away_hu)
            event_date = ev.get("eventDate", "")

            # Find 1X2 market
            market_1x2 = next(
                (m for m in ev.get("markets", []) if m.get("marketName") == "1X2"),
                None
            )
            if not market_1x2:
                continue

            outcomes = {o["outcomeNo"]: o["fixedOdds"] for o in market_1x2.get("outcomes", [])}
            home_odds = outcomes.get(1)
            draw_odds = outcomes.get(2)
            away_odds = outcomes.get(3)

            if not all([home_odds, draw_odds, away_odds]):
                continue

            # Only include if it's a WC competition
            competition = ev.get("competitionName", "")
            is_wc = any(c.get("name") == "vb" for c in ev.get("competitions", []))

            matches.append({
                "home_team_hu": home_hu,
                "away_team_hu": away_hu,
                "home_team": home_en,
                "away_team": away_en,
                "competition": competition,
                "is_wc": is_wc,
                "event_date": event_date,
                "home_odds": home_odds,
                "draw_odds": draw_odds,
                "away_odds": away_odds,
                "event_id": ev.get("eventId"),
            })

    return matches


def _fetch_event_markets(event_id: int) -> list[dict]:
    """Fetch full market groups for a single event (GET /v2/tippmix/event/{id})."""
    try:
        r = requests.get(f"{API_BASE}/v2/tippmix/event/{event_id}",
                         headers=_HEADERS, timeout=20)
        r.raise_for_status()
        return r.json().get("event", {}).get("marketGroups", [])
    except Exception:
        return []


def _fetch_events_list() -> dict | None:
    """POST the list endpoints (best-games, then last-minute) — first 200 wins."""
    for path in ("/tippmix/best-games", "/tippmix/last-minute"):
        try:
            r = requests.post(f"{API_BASE}{path}", headers=_HEADERS, json={}, timeout=20)
            if r.status_code == 200:
                return r.json()
        except Exception:
            continue
    return None


def get_tippmix_matches() -> list[dict]:
    """Returns football matches with Tippmix odds and full market data for WC matches."""
    print("  Fetching Tippmix.hu odds via API...")
    data = _fetch_events_list()
    base_matches = _parse_events(data) if data else []

    # Fetch full market data for WC matches in parallel (keeps us well under
    # Vercel's request time limit even with a dozen WC fixtures).
    wc = [m for m in base_matches if m.get("is_wc") and m.get("event_id")]
    if wc:
        with ThreadPoolExecutor(max_workers=8) as ex:
            groups = list(ex.map(lambda m: _fetch_event_markets(m["event_id"]), wc))
        for m, mgs in zip(wc, groups):
            m["market_groups"] = mgs

    for m in base_matches:
        m.setdefault("market_groups", [])

    other = [m for m in base_matches if not m["is_wc"]]
    print(f"  Got {len(wc)} WC matches + {len(other)} other football matches from Tippmix")
    return base_matches

File: lib/tippmix.ts
/**
 * Tippmix.hu odds fetcher — port of tippmix.py.
 * Calls the public tippmix.hu JSON API directly via fetch (no browser).
 */
import { ProxyAgent } from "undici";
import type { MarketGroup } from "@/lib/valuebets";

const API_BASE = "https://api.tippmix.hu";

// tippmix.hu geo-blocks non-Hungarian IPs. On Vercel (US/EU datacenters) the
// requests are refused, so route them through a Hungarian-exit proxy when
// TIPPMIX_PROXY_URL is set (e.g. "http://user:pass@hu-proxy.example.com:8080").
// Inert locally / when unset — falls back to a direct connection.
const PROXY_URL = process.env.TIPPMIX_PROXY_URL;
const proxyDispatcher = PROXY_URL ? new ProxyAgent(PROXY_URL) : undefined;

/** Merge the proxy dispatcher into fetch options when a proxy is configured. */
function withProxy(init: RequestInit): RequestInit {
  return proxyDispatcher
    ? ({ ...init, dispatcher: proxyDispatcher } as RequestInit)
    : init;
}

const HEADERS: Record<string, string> = {
  Accept: "application/json, text/plain, */*",
  "Content-Type": "application/json;charset=UTF-8",
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko)",
  Referer: "https://www.tippmix.hu/",
  "Accept-Language": "hu-HU",
};

// Hungarian name → English name mapping for WC 2026 teams
export const HU_TO_EN: Record<string, string> = {
  Mexikó: "Mexico",
  "Dél-Afrika": "South Africa",
  Marokkó: "Morocco",
  Portugália: "Portugal",
  Argentína: "Argentina",
  "Szaúd-Arábia": "Saudi Arabia",
  Lengyelország: "Poland",
  Németország: "Germany",
  Japán: "Japan",
  Spanyolország: "Spain",
  "Costa Rica": "Costa Rica",
  Belgium: "Belgium",
  Kanada: "Canada",
  Horvátország: "Croatia",
  Szerbia: "Serbia",
  Brazília: "Brazil",
  Svájc: "Switzerland",
  Kamerun: "Cameroon",
  "Dél-Korea": "Korea Republic",
  Ghána: "Ghana",
  Uruguay: "Uruguay",
  Franciaország: "France",
  Ausztrália: "Australia",
  Dánia: "Denmark",
  Tunézia: "Tunisia",
  Szenegál: "Senegal",
  Hollandia: "Netherlands",
  Ecuador: "Ecuador",
  Anglia: "England",
  Irán: "Iran",
  Wales: "Wales",
  Katar: "Qatar",
  "Egyesült Államok": "United States",
  USA: "United States",
  Panamá: "Panama",
  Jamaika: "Jamaica",
  Bolívia: "Bolivia",
  Venezuela: "Venezuela",
  Kolumbia: "Colombia",
  Chile: "Chile",
  Peru: "Peru",
  Paraguay: "Paraguay",
  Honduras: "Honduras",
  Nigéria: "Nigeria",
  Elefántcsontpart: "Ivory Coast",
  Mali: "Mali",
  Kenya: "Kenya",
  Egyiptom: "Egypt",
  "DR Kongó": "DR Congo",
  Algéria: "Algeria",
  Szlovénia: "Slovenia",
  Ausztria: "Austria",
  Törökország: "Turkey",
  Csehország: "Czech Republic",
  "Új-Zéland": "New Zealand",
  Norvégia: "Norway",
  Svédország: "Sweden",
  Skócia: "Scotland",
  Görögország: "Greece",
  Románia: "Romania",
  Ukrajna: "Ukraine",
  Magyarország: "Hungary",
  Olaszország: "Italy",
  "Bosznia-Hercegovina": "Bosnia & Herzegovina",
  // AFC teams missing from original map
  Üzbegisztán: "Uzbekistan",
  Irak: "Iraq",
  Tadzsikisztán: "Tajikistan",
  Indonézia: "Indonesia",
  Kína: "China",
  Kuvait: "Kuwait",
  Omán: "Oman",
  Bahrain: "Bahrain",
  Jordánia: "Jordan",
  // Other missing teams
  Izland: "Iceland",
  Grúzia: "Georgia",
  Albánia: "Albania",
  "Észak-Macedónia": "North Macedonia",
  Kongó: "Congo",
  Mozambik: "Mozambique",
  Tanzánia: "Tanzania",
  Zambia: "Zambia",
  Etiópia: "Ethiopia",
  Angola: "Angola",
  Finnország: "Finland",
  "Trinidad és Tobago": "Trinidad and Tobago",
  Guatemala: "Guatemala",
  "El Salvador": "El Salvador",
  Guyana: "Guyana",
  Suriname: "Suriname",
};

function en(name: string): string {
  return HU_TO_EN[name] ?? name;
}

export interface TippmixMatch {
  home_team_hu: string;
  away_team_hu: string;
  home_team: string;
  away_team: string;
  competition: string;
  is_wc: boolean;
  event_date?: string;
  home_odds: number | null;
  draw_odds: number | null;
  away_odds: number | null;
  event_id?: number;
  market_groups?: MarketGroup[];
}

function parseEvents(data: any): TippmixMatch[] {
  const matches: TippmixMatch[] = [];
  for (const sportBlock of data?.data ?? []) {
    if (sportBlock.sportId !== 1) continue;
    for (const ev of sportBlock.events ?? []) {
      if (ev.isLive || ev.bettingStatus !== 0) continue;
      const participants = ev.eventParticipants ?? [];
      if (participants.length < 2) continue;

      const homeHu = participants[0].participantName;
      const awayHu = participants[1].participantName;
      const eventDate = ev.eventDate ?? "";

      const market1x2 = (ev.markets ?? []).find(
        (m: any) => m.marketName === "1X2"
      );
      if (!market1x2) continue;

      const outcomes: Record<number, number> = {};
      for (const o of market1x2.outcomes ?? []) outcomes[o.outcomeNo] = o.fixedOdds;
      const homeOdds = outcomes[1];
      const drawOdds = outcomes[2];
      const awayOdds = outcomes[3];
      if (!homeOdds || !drawOdds || !awayOdds) continue;

      const competition = ev.competitionName ?? "";
      const isWc = (ev.competitions ?? []).some((c: any) => c.name === "vb");

      matches.push({
        home_team_hu: homeHu,
        away_team_hu: awayHu,
        home_team: en(homeHu),
        away_team: en(awayHu),
        competition,
        is_wc: isWc,
        event_date: eventDate,
        home_odds: homeOdds,
        draw_odds: drawOdds,
        away_odds: awayOdds,
        event_id: ev.eventId,
      });
    }
  }
  return matches;
}

function parseLiveEvents(data: any): TippmixMatch[] {
  const matches: TippmixMatch[] = [];
  for (const sportBlock of data?.data ?? []) {
    if (sportBlock.sportId !== 1) continue;
    for (const ev of sportBlock.events ?? []) {
      if (!ev.isLive) continue;
      const participants = ev.eventParticipants ?? [];
      if (participants.length < 2) continue;

      const homeHu = participants[0].participantName;
      const awayHu = participants[1].participantName;

      const market1x2 = (ev.markets ?? []).find(
        (m: any) => m.marketName === "1X2"
      );
      let homeOdds: number | null = null;
      let drawOdds: number | null = null;
      let awayOdds: number | null = null;
      if (market1x2) {
        const outcomes: Record<number, number> = {};
        for (const o of market1x2.outcomes ?? [])
          outcomes[o.outcomeNo] = o.fixedOdds;
        homeOdds = outcomes[1] ?? null;
        drawOdds = outcomes[2] ?? null;
        awayOdds = outcomes[3] ?? null;
      }

      const competition = ev.competitionName ?? "";
      const isWc = (ev.competitions ?? []).some((c: any) => c.name === "vb");

      matches.push({
        home_team_hu: homeHu,
        away_team_hu: awayHu,
        home_team: en(homeHu),
        away_team: en(awayHu),
        competition,
        is_wc: isWc,
        home_odds: homeOdds,
        draw_odds: drawOdds,
        away_odds: awayOdds,
        event_id: ev.eventId,
      });
    }
  }
  return matches;
}

async function fetchEventMarkets(eventId: number): Promise<MarketGroup[]> {
  try {
    const r = await fetch(`${API_BASE}/v2/tippmix/event/${eventId}`, withProxy({
      headers: HEADERS,
      cache: "no-store",
      signal: AbortSignal.timeout(20000),
    }));
    if (!r.ok) return [];
    const j = await r.json();
    return j?.event?.marketGroups ?? [];
  } catch {
    return [];
  }
}

async function fetchEventsList(): Promise<any | null> {
  for (const path of ["/tippmix/best-games", "/tippmix/last-minute"]) {
    try {
      const r = await fetch(`${API_BASE}${path}`, withProxy({
        method: "POST",
        headers: HEADERS,
        body: "{}",
        cache: "no-store",
        signal: AbortSignal.timeout(20000),
      }));
      const body = await r.text();
      if (r.status === 200) {
        try {
          return JSON.parse(body);
        } catch {
          // 200 but HTML → a block / anti-bot challenge page, not real data.
          const server = r.headers.get("server") ?? "?";
          const snippet = body.slice(0, 160).replace(/\s+/g, " ");
          console.log(
            `  [tippmix] ${path} -> HTTP 200 non-JSON (server=${server}, ${body.length}b): ${snippet}`
          );
          continue;
        }
      }
      const snippet = body.slice(0, 160).replace(/\s+/g, " ");
      console.log(`  [tippmix] ${path} -> HTTP ${r.status}: ${snippet}`);
    } catch (e) {
      console.log(`  [tippmix] ${path} fetch error: ${e}`);
      continue;
    }
  }
  return null;
}

export async function getTippmixMatches(): Promise<TippmixMatch[]> {
  console.log("  Fetching Tippmix.hu odds via API...");
  const data = await fetchEventsList();
  const baseMatches = data ? parseEvents(data) : [];

  const wc = baseMatches.filter((m) => m.is_wc && m.event_id);
  if (wc.length) {
    const groups = await Promise.all(
      wc.map((m) => fetchEventMarkets(m.event_id as number))
    );
    wc.forEach((m, i) => {
      m.market_groups = groups[i];
    });
  }
  for (const m of baseMatches) m.market_groups ??= [];

  const other = baseMatches.filter((m) => !m.is_wc);
  console.log(
    `  Got ${wc.length} WC matches + ${other.length} other football matches from Tippmix`
  );
  return baseMatches;
}

export async function getTippmixLiveOdds(): Promise<TippmixMatch[]> {
  console.log("  Fetching Tippmix.hu LIVE odds via API...");
  const data = await fetchEventsList();
  const matches = data ? parseLiveEvents(data) : [];

  const wcLive = matches.filter((m) => m.is_wc && m.event_id);
  if (wcLive.length) {
    const groups = await Promise.all(
      wcLive.map((m) => fetchEventMarkets(m.event_id as number))
    );
    wcLive.forEach((m, i) => {
      m.market_groups = Array.isArray(groups[i]) ? groups[i] : [];
    });
  }
  for (const m of matches) m.market_groups ??= [];

  const wc = matches.filter((m) => m.is_wc);
  console.log(
    `  Got ${wc.length} live WC matches from Tippmix (${matches.length} total live)`
  );
  return matches;
}
